ra-tls/libenclave-tls/enclave-tls.c

```diff
@@ -0,0 +1,239 @@
+/* ra-tls-server.c
+ *
+ * Copyright (C) 2006-2015 wolfSSL Inc.
+ *
+ * This file is part of wolfSSL. (formerly known as CyaSSL)
+ *
+ * wolfSSL is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+* (at your option) any later version.
+*
+* wolfSSL is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program; if not, write to the Free Software
+* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+*/
+
+/* the usual suspects */
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <unistd.h>
+
+#include <wolfssl/ssl.h>
+#include <wolfssl/certs_test.h>
+
+#include "enclave-tls.h"
+#include "wolfssl-ra-common.h"
+#include "wolfssl-ra-attester-common.h"
+#include "ra-common.h"
+
+#define DEFAULT_PORT 11111
+#define EPID "epid"
+#define CIPHER_LIST "ECDHE-ECDSA-AES128-GCM-SHA256"
+#include <stdbool.h>
+
+void *handle = NULL;
+
+typedef int (*verify_cert_extensions_func)(uint8_t *, uint32_t);
+
+int cert_verify_callback(int preverify, WOLFSSL_X509_STORE_CTX * store)
+{
+	(void) preverify;
+	int ret = 0;
+	if (is_epid_ratls_cert(store->certs->buffer, store->certs->length)) {
+		void *handle = dlopen("libenclave-tls-sgx.so", RTLD_LAZY);
+		if (NULL == handle) {
+			printf("open error:%s\n", dlerror());
+			return -1;
+		}
+		verify_cert_extensions_func verify_cert_extensions =
+			dlsym(handle, "verify_sgx_cert_extensions");
+		ret = verify_cert_extensions(store->certs->buffer,
+					     store->certs->length);
+		dlclose(handle);
+	}
+
+	fprintf(stderr, "Verifying certificate extensions ... %s\n",
+		ret == 0 ? "Success" : "Failure");
+	return !ret;
+}
+
+WOLFSSL *enclave_tls_init(struct enclave_tls_ctx *e_ctx)
```

Comment:
ctx should be returned rather than WOLFSSL *. Anything returned from enclave_tls API means the caller needs to understand the type and know the definition of type. Actually enclave_tls_ctx should be a non-transparent data type. You need to return ctx using the type `void *`.